### ISSUE TITLE
JSONP callback param is not sanitized

### ISSUE THREAD

**Issue by onebytegone** [silvermine/lambda-express#35]
The [current implementation](https://github.com/silvermine/lambda-express/blob/c7396116de1686f60ac8e7970729ba2ed6e2f610/src/Response.ts#L467-L472) for building a JSONP callback function uses the exact string which was passed to the API. This can create invalid JS if the API was requested with something like: `https://example.com/endpoint?callback=%20not%20valid`.

Looking at [the implementation in express](https://github.com/expressjs/express/blob/3d10279826f59bf68e28995ce423f7bc4d2f11cf/lib/response.js#L322-L323), all but `\[\][a-zA-Z0-9_]` will be filtered from the callback name. Should lambda-express be doing something similar?

For posterity, other express items that might be of use:

* https://github.com/expressjs/express/commit/0ae18bca6075c8f6d9a7d58405c5f52c1ae61958#commitcomment-465075
* https://github.com/expressjs/express/pull/502/files
* https://github.com/expressjs/express/commit/df2584cc3bca2829eeb2799df910ce27e317f67e
* https://github.com/expressjs/express/commit/89c5affc3ba24f5186feb493cb4db9dc473f3479

**Comment by jthomerson** [comment]
Yeah, we should do this. Ironically, we did it in our older and clunkier [API Gateway Utils](https://github.com/silvermine/apigateway-utils/blob/251419bd69d0d9419ef86cb83c29fbfbf4944e28/src/ResponseBuilder.js#L245).

Interestingly, the express implementation does not check again to see if the string is non-empty after it has removed invalid characters. I think (although I haven't tested it) this would mean that a callback of entirely invalid characters would result in invalid JS being written. We should be sure that doesn't happen for us.

I'm also not entirely sure I like the express method of just removing the invalid characters and then proceeding with the JSONP response. The question is: under what circumstances would the callback have invalid characters in it? Likely either user manipulation / attack, or programmer error. If it's user manipulation, you really don't want to continue executing the JSONP code, right? And if it's programmer error, who does it help to make a JSONP callback to a function that's not defined because you modified the name of the function that you're calling? (Yes, I realize the undefined check will keep the invocation from happening, but the result is the same - nothing was called).

Thus, let's take an approach that mixes what Express did and what we did in APIGW utils:

 * check if the callback is an array; if so, use the first value
 * if that value (callback name) is a *valid string*, invoke JSONP
    * if it's *invalid*, just continue with regular JSON response

Let me know if that's not clear.

**Comment by onebytegone** [comment]
@jthomerson Makes sense to me. As for merging behaviors, I think `/^[\[\]\w$.]+$/` is the "is valid?" regex we want to use. It should be the "only allow these chars" version of what is [used in express](https://github.com/expressjs/express/blob/3d10279826f59bf68e28995ce423f7bc4d2f11cf/lib/response.js#L322-L323) (i.e. `/[^\[\]\w$.]/`).  However, as seen from the last 3 tests tests below, this pattern will create a slightly different behavior from what is in API Gateway Utils. Thoughts?

```
$ node
> function check(str) {
... return {
..... apigwUtils: /^[A-Za-z0-9_\\$]+$/.test(str),
..... express: /^[\[\]\w$.]+$/.test(str),
..... };
... };
undefined
> check('callback');
{ apigwUtils: true, express: true }
> check('_abc');
{ apigwUtils: true, express: true }
> check('$');
{ apigwUtils: true, express: true }
> check('call-back');
{ apigwUtils: false, express: false }
> check('bad;func()');
{ apigwUtils: false, express: false }
> check('fn[123]');
{ apigwUtils: false, express: true }
> check('window.document');
{ apigwUtils: false, express: true }
> check('some\\func');
{ apigwUtils: true, express: false }
```

**Comment by jthomerson** [comment]
Thanks for the thorough testing! Let's go with what Express uses.

**Comment by jthomerson** [comment]
To be clear, let's go with your proposal for the "is valid?" check, despite the fact that the behavior will differ from APIGW utils.

**Comment by onebytegone** [comment]
Closed with #42.
